v2/internal/controller/controllers/targetnamespace_test.go

```diff
@@ -0,0 +1,258 @@
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	resources "github.com/Azure/azure-service-operator/v2/api/microsoft.resources/v1alpha1api20200601"
+	"github.com/Azure/azure-service-operator/v2/internal/controller/controllers"
+	"github.com/Azure/azure-service-operator/v2/internal/controller/reconcilers"
+	"github.com/Azure/azure-service-operator/v2/internal/controller/testcommon"
+
+	. "github.com/onsi/gomega"
+)
+
+const finalizerName = reconcilers.GenericControllerFinalizer
+
+const (
+	timeoutFast = time.Minute * 3
+	retry       = time.Second * 3
+)
+
+func TestTargetNamespaces(t *testing.T) {
+	t.Parallel()
+	tc := globalTestContext.ForTest(t)
+
+	createNamespaces(tc, "watched", "unwatched")
+	configuredNamespaces := os.Getenv("AZURE_TARGET_NAMESPACES")
+	podNamespace := os.Getenv("POD_NAMESPACE")
+
+	// We can't check for operator namespace if there's not one set.
+	tc.Expect(podNamespace).ToNot(Equal(""), "no POD_NAMESPACE set")
+
+	standardSpec := resources.ResourceGroupSpec{
+		Location: tc.AzureRegion,
+		Tags:     testcommon.CreateTestResourceGroupDefaultTags(),
+	}
+	// Create resource groups in these two namespaces - we can't
+	// easily use the test context for this because that's geared to
+	// creating things in one test namespace, while we need fixed
+	// namespaces.
+	rgDefault := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tc.Namer.GenerateName("rg"),
+			Namespace: "default",
+		},
+		Spec: standardSpec,
+	}
+	tc.CreateResourceGroupAndWait(&rgDefault)
+	// Check that the instance is annotated with the operator namespace.
+	checkNamespaceAnnotation(tc, &rgDefault, podNamespace)
+
+	// The watched namespace is also reconciled.
+	rgWatched := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tc.Namer.GenerateName("rg"),
+			Namespace: "watched",
+		},
+		Spec: standardSpec,
+	}
+	tc.CreateResourceGroupAndWait(&rgWatched)
+	checkNamespaceAnnotation(tc, &rgWatched, podNamespace)
+
+	// But the unwatched namespace isn't...
+	unwatchedName := tc.Namer.GenerateName("rg")
+	rgUnwatched := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      unwatchedName,
+			Namespace: "unwatched",
+		},
+		Spec: standardSpec,
+	}
+	_, err := tc.CreateResourceGroup(&rgUnwatched)
+	tc.Expect(err).ToNot(HaveOccurred())
+
+	name := types.NamespacedName{Name: unwatchedName, Namespace: "unwatched"}
+
+	gotNamespaceAnnotation := func(g Gomega) bool {
+		var instance resources.ResourceGroup
+		ctx := context.Background()
+		err := tc.KubeClient.Get(ctx, name, &instance)
+		g.Expect(err).NotTo(HaveOccurred())
+		res, err := meta.Accessor(&instance)
+		g.Expect(err).NotTo(HaveOccurred())
+		return res.GetAnnotations()[controllers.NamespaceAnnotation] == podNamespace
+	}
+
+	if configuredNamespaces == "" {
+		t.Log("**** all namespaces mode")
+		// The operator should be watching all namespaces.
+		tc.G.Eventually(
```

Comment:
Actually you should just use `tc.Eventually` and then you don't need to fuss with `RemainingTime()` and `PollingInterval()` that I mentioned above here, although you might need to for the `Consistently` below...